Python/Product/Analysis/Interpreter/Ast/AstPythonInterpreterFactory.cs

```diff
@@ -643,32 +672,67 @@ TryImportModuleContext context
 
             var mp = mmp.Value;
 
-#if USE_TYPESHED
-            lock (_typeShedPathsLock) {
-                if (_typeShedPaths == null) {
-                    var typeshed = FindModuleInSearchPath(_factory.GetSearchPaths(), _factory.GetImportableModules(), "typeshed");
-                    if (typeshed.HasValue) {
-                        _typeShedPaths = GetTypeShedPaths(PathUtils.GetParent(typeshed.Value.SourceFile)).ToArray();
-                    } else {
-                        _typeShedPaths = Array.Empty<string>();
-                    }
-                }
-                if (_typeShedPaths.Any()) {
-                    var mtsp = FindModuleInSearchPath(_typeShedPaths, null, mp.FullName);
-                    if (mtsp.HasValue) {
-                        mp = mtsp.Value;
-                    }
-                }
-            }
-#endif
+            IPythonModule module;
 
             if (mp.IsCompiled) {
                 _log?.Log(TraceLevel.Verbose, "ImportScraped", mp.FullName, FastRelativePath(mp.SourceFile));
-                return new AstScrapedPythonModule(mp.FullName, mp.SourceFile);
+                module = new AstScrapedPythonModule(mp.FullName, mp.SourceFile);
+            } else {
+                _log?.Log(TraceLevel.Verbose, "Import", mp.FullName, FastRelativePath(mp.SourceFile));
+                module = PythonModuleLoader.FromFile(context.Interpreter, mp.SourceFile, LanguageVersion, mp.FullName);
+            }
+
+            return module;
+        }
+
+        private IReadOnlyList<string> GetTypeShedPaths(int timeout) {
+            // First check
+            var typeShedPaths = _typeShedPaths;
```

Comment:
`_typeShedPaths` is accessed outside of the lock

Reply:
Deliberately, so we can avoid locking for the search if the search has already been done (it'll only ever be done once - we never change `_typeShedPaths` after initialization - but it depends on context too much to make it `Lazy<T>`).

Reply:
Since `_searchPathsLock` is used only for locking, it will be a thin lock and the overhead of using it is close to nothing:
https://blogs.msdn.microsoft.com/seteplia/2017/09/06/managed-object-internals-part-2-object-header-layout-and-the-cost-of-locking/
http://firstclassthoughts.co.uk/Articles/Programming/DotNetLocksAreReallyFast.html

Reply:
Not using it has no overhead, and it's only for initialization.

Reply:
It's also read outside the `async` to avoid any overhead there.